### PR TITLE
feat: add switch

### PR DESCRIPTION
**Pull request by jose-costa-frontify**
I have some doubts about the design and how the indeterminate state should work.

Clickup: [8694gat8g](https://app.clickup.com/t/8694gat8g)
Figma: [Switch](https://www.figma.com/design/WfEbqZI0nnPWHN43MCtvnm/Fondue---Inputs?node-id=569-83144&m=dev)

## Definition of Done
- [x] User interface and experience have been verified by a designer
- [x] I have added thorough Cypress tests (All properties and interactions have been tested).
- [x] Cross-browser compatibility - The component look consistent in the latest version of Chrome, Safari, Firefox and Edge.
- [x] I've added documentation in Storybook. all properties are reflected in table and controls.
- [x] User interface is compliant with WCAG 2.1 AA. Ensure these items are fulfilled:
    - [x] Keyboard operability: All functionality can be accessed using only the keyboard
    - [x] Logical tab order: Interactive elements follow a consistent and logical tab order
    - [x] Visual indication of focus: Focused elements are visually highlighted for keyboard navigation
    - [x] Color contrast: Sufficient contrast is used between text and background, as well as form fields and background
    - [x] Alternative visual cues: Information is not solely reliant on color and includes alternative visual indicators
    - [x] Semantic markup: Headings, lists, and form elements are marked up with appropriate semantic tags

### CODE REVIEW

File: packages/components/MIGRATING.md [hunk 1]
```diff
@@ -28,12 +28,15 @@ This document describes the changes that you need to make to your code to migrat
...
+        -   [Switch](#switch)
             -   [Old](#old-7)
             -   [New](#new-7)
-        -   [Tooltip](#tooltip)
+        -   [Text Input](#text-input)
             -   [Old](#old-8)
             -   [New](#new-8)
+        -   [Tooltip](#tooltip)
+            -   [Old](#old-9)
+            -   [New](#new-0)
```
Comment (SamuelAlev) [inline packages/components/MIGRATING.md:39]:
```suggestion
            -   [New](#new-9)
```

File: packages/components/MIGRATING.md [hunk 2]
```diff
@@ -434,6 +437,76 @@ Changes:
...
+    -   `aria-describedby` has been added.
+
+-   **Event Handling**:
+
+    -   The `onChange` event now returns the latest state of the value rather than a `MouseEvent`.
+    -   `onBlur` and `onFocus` callbacks have been added.
+
+-   **Removed Properties**:
+
+    -   `label`, `labelStyle` and `tooltip` have been removed in favour of composition.
```
Comment (SamuelAlev) [inline packages/components/MIGRATING.md:463]: We use American English
```suggestion
    -   `label`, `labelStyle` and `tooltip` have been removed in favor of composition.
```

File: packages/components/MIGRATING.md [hunk 3]
```diff
@@ -434,6 +437,76 @@ Changes:
...
+
+```tsx
+<Tooltip.Root enterDelay={200}>
+    <Tooltip.Trigger>
+        <Label id="dark-mode-label" htmlFor="new-switch">Dark mode</Label>
+    </Tooltip.Trigger>
+    <Tooltip.Content side="top" padding="spacious">
+        <p id="dark-mode-switch-description">Toggle dark mode.</p>
+    </Tooltip.Content>
+</Tooltip.Root>
```
Comment (SamuelAlev) [inline packages/components/MIGRATING.md:492]: This part of the example is wrong `tooltip` added a question-mark-circle icon with the tooltip text
Screenshot 2024-07-23 at 13 34 40

File: packages/components/MIGRATING.md [hunk 4]
```diff
@@ -434,6 +437,76 @@ Changes:
...
+<Switch
+    id="new-switch"
+    name="dark-mode-switch"
+    defaultValue={true}
+    size="medium"
+    hug={true}
```
Comment (SamuelAlev) [inline packages/components/MIGRATING.md:498]: Why would a switch be full width?

File: packages/components/src/components/Switch/Switch.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+import { type CommonAriaAttrs } from '#/utilities/types';
+
+import styles from './styles/switch.module.scss';
+
+type SwitchProps = {
+    id?: string;
+    name?: string;
+    /**
+     * The size of the switch component.
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/Switch.tsx:14]: Please add default values

File: packages/components/src/components/Switch/Switch.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        size = 'medium',
+        hug = false,
+        onChange,
+        'data-test-id': dataTestId = 'fondue-switch',
+        'aria-label': ariaLabel = 'Switch',
+        ...props
+    }: SwitchProps,
+    ref: ForwardedRef<HTMLButtonElement>,
+) => {
+    const rootClasses = [styles.root, styles[size], hug && styles.fullwidth].join(' ');
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/Switch.tsx:65]:
```suggestion
    const rootClasses = cn(styles.root, styles[size], hug && styles.fullwidth);
```

**Review by SamuelAlev — CHANGES_REQUESTED** [review]

File: packages/components/src/components/Switch/styles/switch.module.scss [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            position: absolute;
+            top: 50%;
+            left: 50%;
+            transform: translate(-50%, -50%);
+            width: 30%;
+            height: 10%;
+            background-color: var(--base-color);
+            border-radius: 9999px;
+        }
+    }
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/styles/switch.module.scss:40]: Can be removed

File: packages/components/src/components/Switch/styles/switch.module.scss
Comment (SamuelAlev) [inline packages/components/src/components/Switch/styles/switch.module.scss:1]: missing disabled state styles

File: packages/components/MIGRATING.md (hunk 4 above)
Comment (jose-costa-frontify) [reply packages/components/MIGRATING.md:498]: indeed

File: packages/components/src/components/Switch/Switch.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        size: 'medium',
+    },
+};
+
+export const SizeLarge: Story = {
+    args: {
+        defaultValue: false,
+        size: 'large',
+    },
+};
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/Switch.stories.tsx:88]: Can you merge them into 1 story?

File: packages/components/src/components/Switch/Switch.tsx (hunk 6 above)
Comment (jose-costa-frontify) [reply packages/components/src/components/Switch/Switch.tsx:65]: thought this was only for tailwind classes on second look this is to extend with the given classes tailwind?

File: packages/components/src/components/Switch/styles/switch.module.scss
Comment (jose-costa-frontify) [reply packages/components/src/components/Switch/styles/switch.module.scss:1]: thanks, completely escaped me

File: packages/components/MIGRATING.md [hunk 9]
```diff
@@ -434,6 +437,76 @@ Changes:
...
+    -   `aria-describedby` has been added.
+
+-   **Event Handling**:
+
+    -   The `onChange` event now returns the latest state of the value rather than a `MouseEvent`.
+    -   `onBlur` and `onFocus` callbacks have been added.
+
+-   **Removed Properties**:
+
+    -   `label`, `labelStyle` and `tooltip` have been removed in favor of composition.
```
Comment (SamuelAlev) [inline packages/components/MIGRATING.md:463]: + `hug` in removed properties

File: packages/components/src/components/Switch/styles/switch.module.scss [hunk 10]
```diff
@@ -0,0 +1,61 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+@import '../../../utilities/sizeToken.module.scss';
+@import '../../../utilities/focusStyle.module.scss';
+@import '../../../utilities/transitions.module.scss';
+
+.root {
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/styles/switch.module.scss:7]:
```suggestion
.switch {
```

File: packages/components/src/components/Switch/__tests__/Switch.ct.tsx [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+test('should apply size and full width classes', async ({ mount }) => {
+    const component = await mount(<Switch aria-label={ARIA_LABEL} />);
+    await expect(component).toHaveClass(/medium/);
+
+    await component.update(<Switch aria-label={ARIA_LABEL} size="large" />);
+    await expect(component).toHaveClass(/large/);
+
+    await component.update(<Switch aria-label={ARIA_LABEL} size="small" />);
+    await expect(component).toHaveClass(/small/);
+});
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/__tests__/Switch.ct.tsx:44]:
```suggestion
test('should apply size classes', async ({ mount }) => {
    const component = await mount(<Switch aria-label={ARIA_LABEL} />);
    await expect(component).toHaveClass(/medium/);

    await component.update(<Switch aria-label={ARIA_LABEL} size="large" />);
    await expect(component).toHaveClass(/large/);

    await component.update(<Switch aria-label={ARIA_LABEL} size="small" />);
    await expect(component).toHaveClass(/small/);
});
```

File: .changeset/chatty-plums-kiss.md [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
+"@frontify/fondue-components": minor
+---
+
+feat: add switch
```
Comment (SamuelAlev) [inline .changeset/chatty-plums-kiss.md:5]:
```suggestion
feat: add `Switch` component
```

File: .changeset/chatty-plums-kiss.md
Comment (SamuelAlev) [inline .changeset/chatty-plums-kiss.md:1]: Could you mark the old Switch component as deprecated?

Comment (SamuelAlev) [inline .changeset/chatty-plums-kiss.md:1]: Could you add a deprecation notice on the Story file like done here https://github.com/Frontify/fondue/blob/f56adc12c183d9101aa35002df75add674f9ff00/packages/fondue/src/components/Button/Button.stories.tsx#L24

(must be on top of the export default in the story)

File: packages/components/src/components/Switch/Switch.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+        return <Switch value={checked} onChange={onChange} {...args} />;
+    },
+};
+
+export const Size: Story = {
+    args: {
+        defaultValue: false,
+        size: 'small',
+    },
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/Switch.stories.tsx:71]: Could also be a rendering of the 3 sizes:
```suggestion
    render: (args) => (
        <>
            <Switch {...args} size="small" />
            <Switch {...args} size="medium" />
            <Switch {...args} size="large" />   
        </>
    ),
```

File: packages/components/MIGRATING.md [hunk 14]
```diff
@@ -434,6 +437,76 @@ Changes:
...
+    -   `aria-describedby` has been added.
+
+-   **Event Handling**:
+
+    -   The `onChange` event now returns the latest state of the value rather than a `MouseEvent`.
+    -   `onBlur` and `onFocus` callbacks have been added.
+
+-   **Removed Properties**:
+
+    -   `label`, `labelStyle`, `hug`, and `tooltip` have been removed in favor of composition.
```
Comment (SamuelAlev) [inline packages/components/MIGRATING.md:463]:
```suggestion
    -   `label`, `labelStyle` and `tooltip` have been removed in favor of composition.
    -   `hug` have been removed without replacement
```

File: packages/components/src/components/Switch/Switch.tsx [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    ref: ForwardedRef<HTMLButtonElement>,
+) => {
+    const rootClasses = cn([styles.switch, styles[size]]);
+
+    return (
+        <RadixSwitch.Root
+            ref={ref}
+            checked={value}
+            defaultChecked={defaultValue}
+            className={rootClasses}
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/Switch.tsx:76]:
```suggestion
    return (
        <RadixSwitch.Root
            ref={ref}
            checked={value}
            defaultChecked={defaultValue}
            className={cn([styles.switch, styles[size]])}
```

File: packages/fondue/src/components/Switch/Switch.tsx [hunk 16]
```diff
@@ -49,7 +49,9 @@ export type SwitchProps = {
     onChange?: (e: MouseEvent) => void;
     'data-test-id'?: string;
 };
-
+/**
```
Comment (SamuelAlev) [inline packages/fondue/src/components/Switch/Switch.tsx:52]:
```suggestion

/**
```

File: packages/components/src/components/Switch/styles/switch.module.scss [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        display: block;
+        background-color: var(--base-color);
+        border-radius: 9999px;
+        border: 1px solid var(--box-neutral-inverse-color);
+        transition: transform 100ms;
+        transform: translate(-1px, -1px);
+        will-change: transform;
+        width: calc(100cqh + 2px);
+        height: calc(100cqh + 2px);
+        &[data-state='checked'] {
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/styles/switch.module.scss:61]:
```suggestion
        height: calc(100cqh + 2px);

        &[data-state='checked'] {
```

File: packages/components/src/components/Switch/styles/switch.module.scss [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+    &.large {
+        width: sizeToken(9);
+        height: sizeToken(5);
+    }
+
+    &[disabled] {
+        border-color: var(--line-color);
+        background-color: var(--box-neutral-color);
+        .thumb {
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/styles/switch.module.scss:46]:
```suggestion
        background-color: var(--box-neutral-color);

        .thumb {
```

File: packages/components/src/components/Switch/styles/switch.module.scss [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    border: 1px solid var(--line-color-x-strong);
+    container-type: size;
+
+    &:hover {
+        background-color: var(--box-neutral-color-hover);
+    }
+
+    &[data-state='checked'] {
+        background-color: var(--text-color-weak);
+        &:hover {
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/styles/switch.module.scss:23]:
```suggestion
        background-color: var(--text-color-weak);

        &:hover {
```

File: packages/components/src/components/Switch/Switch.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+};
+
+export const Size: Story = {
+    render: (args) => (
+        <div style={{ display: 'flex', gap: 16, alignItems: 'center' }}>
+            <Switch {...args} size="small" />
+            <Switch {...args} size="medium" />
+            <Switch {...args} size="large" />
+        </div>
+    ),
```
Comment (SamuelAlev) [inline packages/components/src/components/Switch/Switch.stories.tsx:106]: Tip: use `decorator` so you have a clean "show code"
Currently, we have a wrapping div with style shown, which is noise to the dev:
Screenshot 2024-07-23 at 18 21 09

```suggestion
    decorators: [
        (Story) => {
            return (
                <div className="tw-flex tw-flex-col tw-gap-4 tw-items-center">
                    <Story />
                </div>
            );
        },
    ],
    render: (args) => (
        <>
            <Switch {...args} size="small" />
            <Switch {...args} size="medium" />
            <Switch {...args} size="large" />
        </>
    ),
```

**Review by SamuelAlev — APPROVED** [review]
It looks good!

**Pull request merged**